contracts/state/StateV2.sol

```diff
@@ -0,0 +1,464 @@
+            );
+        }
+        // Binary search
+        uint256 min = 0;
+        uint256 max = identities[id].length - 1;
+        while (min <= max) {
+            uint256 mid = (max + min) / 2;
+            if (identities[id][mid].BlockTimestamp == timestamp) {
+                return (
+                    identities[id][mid].BlockN,
+                    identities[id][mid].BlockTimestamp,
+                    identities[id][mid].State
+                );
+            } else if (
+                (timestamp > identities[id][mid].BlockTimestamp) &&
+                (timestamp < identities[id][mid + 1].BlockTimestamp)
+            ) {
+                return (
+                    identities[id][mid].BlockN,
+                    identities[id][mid].BlockTimestamp,
+                    identities[id][mid].State
+                );
+            } else if (timestamp > identities[id][mid].BlockTimestamp) {
+                min = mid + 1;
+            } else {
+                max = mid - 1;
+            }
+        }
+        return (0, 0, 0);
+    }
+
+    /**
+     * @dev Retrieve identity last committed information
+     * @param id identity
+     * @return last state for a given identity
+     * return parameters are (by order): block number, block timestamp, state
+     */
+    function getStateDataById(uint256 id)
+        public
+        view
+        returns (
+            uint64,
+            uint64,
+            uint256
+        )
+    {
+        if (identities[id].length == 0) {
+            return (0, 0, 0);
+        }
+        IDState memory lastIdState = identities[id][identities[id].length - 1];
+        return (
+            lastIdState.BlockN,
+            lastIdState.BlockTimestamp,
+            lastIdState.State
+        );
+    }
+
+    function getSmtCurrentRoot() public view returns (uint256) {
+        return smt.root();
+    }
+
+    function getSmtProof(uint256 _index)
+        public
+        view
+        returns (Proof memory)
+    {
+        return smt.getProof(_index);
+    }
+
+    function getSmtHistoricalProofByRoot(uint256 index, uint256 _root)
+    public
+    view
+    returns (Proof memory)
+    {
+        return smt.getHistoricalProofByRoot(index, _root);
+    }
+
+    function getSmtHistoricalProofByBlock(uint256 index, uint64 _block)
+        public
+        view
+        returns (Proof memory)
+    {
+        return smt.getHistoricalProofByBlock(index, _block);
+    }
+
+    function getSmtHistoricalProofByTime(uint256 index, uint64 timestamp)
+        public
+        view
+        returns (Proof memory)
+    {
+        return smt.getHistoricalProofByTime(index, timestamp);
+    }
+
+    function migrateStateToSmt(
+        uint256 id,
+        uint256 state,
+        uint64 timestamp,
+        uint64 blockNumber
+    ) public onlyOwner {
+        require(!_stateTransitionEnabled, "smt migration is not allowed");
+        smt.addHistorical(id, state, timestamp, blockNumber);
```

Comment:
should we check the result of the function call explicitly?

Reply:
Not sure I understand what you mean.